Upload CSVs

ghickman opened the pull request:
This switches CSV importing from the CLI to a web form.

The importing itself it done by a Queue worker since it takes over 10 seconds.

The uploaded files are stored on the Version created when the form is submitted and we store a processed timestamp so we can infer the created but unprocessed state.

ghickman committed:
Add fields to Version to track CSVs and processing time

ghickman committed:
Add a single entry point for importers based on a given Version

ghickman committed:
Add a page to upload the CSV files

Since importing CSV files isn't fast (10-30s minimum) we create the
Version, save the files to that Version, and then put a job on a queue
(with RQ) to be processed later.

ghickman committed:
Optimise the importers to avoid get_or_create in a loop

ghickman committed:
Conditionally save files to S3

ghickman committed:
Grab Version creators for Version list

ghickman committed:
Update readme to better explain how this project is laid out

fredkingham commented on ncdr/importers/grouping.py:
```diff
@@ -1,24 +1,30 @@
+    fd.readline()  # ignore the first line since it's blank
+    rows = list(csv.DictReader(fd, delimiter="¬"))
+
+    existing_names = {g.name: g.description for g in Grouping.objects.all()}
```
It looks like we're only using the name from the existing names. Can we not do

```
set(Grouping.objects.all().values_list('name', flat=True))
```

Without trying it, I'm assuming this is faster and simpler. Note "Without trying it"!

ghickman replied:
Oops, that's some left over from an unnecessary change, good catch!  Fixed and rebased.

fredkingham commented on ncdr/static/css/styles.scss:
```diff
@@ -657,3 +657,21 @@ hr {
+  }
+
+  input[type="file"] {
+    display: flex;
```
Have we checked on IE? Browser support for flex is not 100% but I'm not sure how limited.

ghickman replied:
Sort of? https://caniuse.com/#feat=flexbox

Do you have an easy way to test this?

fredkingham commented on ncdr/importers/column.py:
```diff
@@ -1,11 +1,34 @@
+    This function builds that table, creating any missing DataElements in the
+    process.
+    """
+    existing_de_names = set(DataElement.objects.values_list("name", flat=True))
```
we could do existing_de_names = set(DataElement.objects.all().filter(name__in= csv_de_names).values_list("name", flat=True))

(bit bike sheddy...)

ghickman replied:
You still have to compute `missing_de_names` though so you've asked the database to do a slightly harder query (by adding a filter clause) without saving the CPU any time when it calculates the set difference.

fredkingham commented:
Should we add pre-commit to `requirments-dev.txt`?

fredkingham commented:
Can we remove load date from the readme

fredkingham commented:
Don't we need to add the redis install to ansible?

fredkingham commented:
Can we create an initial version?

I tried migrating existing data and migrations blew up. I nuked the db and created and I get 

```

Version matching query does not exist.
--

```

(I could be being stupid!)

ghickman committed:
Pin dev requirements so CI runs the same versions as local

ghickman committed:
Add and pin pre-commit hook for local dev

ghickman commented:
> Should we add pre-commit to requirments-dev.txt?

oh go on then ;) 

> Can we remove load date from the readme

Done and rebased.


> Don't we need to add the redis install to ansible?

Yes, yes we do.  Added and rebased in.

ghickman committed:
Show when a version was processed in the list

ghickman committed:
Hash uploaded files to avoid versions with duplicate [aggregate] files

ghickman committed:
Disable [un]publish buttons when a Version hasn't been processed

ghickman committed:
Send RQ exception logs to support email address

ghickman committed:
Generate DataElement and Grouping slugs when they're created

These two models are only created in the importer scripts.  Rather than
hide the slugification of their names in their save()s it's more
explicit to create them appropriate slugs.  They're also now created
with bulk_create which does not call the save()s.

ghickman committed:
Create an initial Version so the site works before uploading files

ghickman commented:
@fredkingham – updated to add an initial version and fix some errors I found when running from a clean database.  You should be able to review locally now!

ghickman committed:
Centralise log output to respective stderr & stdout files

This removes the gunicorn config file since it was both a source of
configuration indirection and didn't allow the use of hyphen to redirect
to standard streams.  As part of that it also consolidates all output
from Gunicorn, Django, and the RQ worker into the same streams.

gunicorn.log is no longer created since logfile is synonymous with
error-logfile to gunicorn and we now redirect that to stderr.

ghickman committed:
Run the RQ worker with circus

ghickman committed:
Don't try to load_data now we upload via the web form

fredkingham commented:
When we're logged in can we have a button in the logged in bar to take you to `/upload`

ghickman commented:
@fredkingham – sure.  Should it take you to `/upload` or `/versions` where upload is also linked?

fredkingham commented:
@ghickman `/version`s is a better idea! (I chose looked for the url that would let me upload 😃 )

fredkingham commented:
I've done my upload

![Screen Shot 2019-03-12 at 16 32 26](https://user-images.githubusercontent.com/2175455/54218202-0397f280-44e5-11e9-861b-b12198a6d267.png)

fredkingham commented:
<img width="1353" alt="Screen Shot 2019-03-13 at 10 41 18" src="https://user-images.githubusercontent.com/2175455/54272929-99329100-457c-11e9-80ea-ff3c554fd0f7.png">

We're missing a distinct somewhere maybe?

Steps to reproduce, upload files. Reupload files with a table name slightly changed.
Search for the database that contains that table.

We now see two databases in the matches, but both lead to the same url.

ghickman committed:
Only search for models related to the current version

Instead of searching for any published version (for which there could be
many) trust the currently set version in middleware.  This also reflect
a version an admin has explicitly set on their request.user.

ghickman committed:
Don't create a version if anything breaks during upload

This specifically avoids the case where Redis can't be contacted.

ghickman committed:
Unpublish all other versions when publishing another

fredkingham commented:
I'm getting a 500 when I try and publish the new version 
<img width="1172" alt="Screen Shot 2019-03-13 at 11 42 03" src="https://user-images.githubusercontent.com/2175455/54276356-106c2300-4585-11e9-9209-bed1a2fdcb09.png">

ghickman committed:
Reverse the auth urls to take into account being deployed under a URL

ghickman committed:
Add a link to the versions page to the logged in banner

ghickman commented:
@fredkingham – fixed the publishing bug, forgot I'd built `Version.[un]publish()` to handle audit logs already!  I've also fixed a bug with auth URL redirects and added that versions link to the logged in header.

ghickman committed:
Add favicon

ghickman committed:
Set an upload limit bigger than the files being uploaded

ghickman committed:
Redirect root URL favicon requests to the application

ghickman committed:
Install boto in production

ghickman committed:
Set the missing bucket name setting

ghickman committed:
Add env var templates for AWS creds

ghickman committed:
Fix Groupings menu highlight

fredkingham commented:
I'm getting errors in the queue

`Version matching query does not exist`

```
NCDR-reference-library/ncdr/importers/__init__.py", line 32, in import_data
    version = Version.objects.get(pk=version_pk)
```

fredkingham commented:
I'm still blowing up on version upload with Version does not exist. 

Initially it was because I had multiple versions published (I think). Now I'm not sure...

ghickman committed:
Log out a string when processing fails

exception() expects an event string, not an exception instance but will
add `exc_info=True` under the hood to capture the traceback for us.

ghickman committed:
Reraise exceptions encountered while processing to trigger a ROLLBACK

ghickman committed:
Don't use IntegrityError to check for existing file hashes

Since this code will be run inside a transaction an IntegrityError will
cause the transaction to fail even if its caught and dealt with.  Since
were using that exception to handle a set of files already existing we
can replace it with a get().

fredkingham commented:
👍

ghickman committed:
Catch an existing version not existing
